### PR TITLE
Added TextCNN wrapper TorchModel

### PR DESCRIPTION
**Pull request by Shiva-sankaran**
## Description

Following up on #3828. Second PR for #3828 , adds TextCNN wrapper TorchModel.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Review by rbharath — COMMENTED** [review]
Can you please review @shreyasvinaya

File: deepchem/models/torch_models/text_cnn.py [hunk 1]
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
             output = output.view(-1, self.n_tasks, 1)
             outputs = [output]
         return outputs
+
+
+class TextCNNModel(TorchModel):
```
Comment (rbharath) [inline deepchem/models/torch_models/text_cnn.py:191]: Let's add this to the docs/

Comment (Shiva-sankaran) [reply deepchem/models/torch_models/text_cnn.py:191]: @shreyasvinaya I have a small doubt regarding updating docs. There is already an entry in the documentation for the TextCNNModel(https://deepchem.readthedocs.io/en/latest/api_reference/models.html#textcnnmodel), which currently links back to the TF implementation, should I replace it with this Torch version.

Comment (Shiva-sankaran) [reply deepchem/models/torch_models/text_cnn.py:191]: Also do I replace Keras with PyTorch in this table(https://deepchem.readthedocs.io/en/latest/api_reference/models.html#id134) as well? I don't see any entries of Keras/PyTorch

Comment (shreyasvinaya) [reply deepchem/models/torch_models/text_cnn.py:191]: add this to the docs under `torch_models`

Comment (Shiva-sankaran) [reply deepchem/models/torch_models/text_cnn.py:191]: Added to docs.

File: deepchem/models/torch_models/text_cnn.py [hunk 2]
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
...
+    # Below functions were taken from DeepChem TextCNN tensorflow implementation
+
+    def default_generator(
+            self,
+            dataset: Dataset,
+            epochs: int = 1,
+            mode: str = 'fit',
+            deterministic: bool = True,
+            pad_batches: bool = True) -> Iterable[Tuple[List, List, List]]:
+        """Transfer smiles strings to fixed length integer vectors"""
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/text_cnn.py:301]: Documentation on parameters to be added

File: deepchem/models/torch_models/text_cnn.py [hunk 3]
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
...
+                # Transform SMILES sequence to integers
+
+                X_b = self.smiles_to_seq_batch(ids_b)
+                yield ([X_b], [y_b], [w_b])
+
+    @staticmethod
+    def build_char_dict(dataset: Dataset,
+                        default_dict: Dict[str, int] = default_dict):
+        """ Collect all unique characters(in smiles) from the dataset.
+       This method should be called before defining the model to build appropriate char_dict
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/text_cnn.py:320]: Documentation on parameters to be added

File: deepchem/models/torch_models/text_cnn.py [hunk 4]
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
...
+                i = i + 1
+        # Add all extra_keys to char_dict
+        for extra_key in extra_keys:
+            out_dict[extra_key] = current_key_val
+            current_key_val += 1
+        return out_dict, seq_length
+
+    def smiles_to_seq(self, smiles: str):
+        """
+        Tokenize characters in smiles to integers
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/text_cnn.py:355]: Documentation on parameters to be added

File: deepchem/models/torch_models/text_cnn.py [hunk 5]
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
...
+                i = i + 1
+            else:
+                raise ValueError('character not found in dict')
+        for i in range(self.seq_length - len(seq)):
+            # Padding with '_'
+            seq.append(self.char_dict['_'])
+        return np.array(seq, dtype=np.int32)
+
+    @staticmethod
+    def convert_bytes_to_char(s):
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/text_cnn.py:380]: Documentation on parameters to be added
type annotations missing

File: deepchem/models/torch_models/text_cnn.py [hunk 6]
```diff
@@ -179,3 +186,210 @@ def forward(self, input: OneOrMany[torch.Tensor]) -> List[Any]:
...
+        s = ''.join(chr(b) for b in s)
+        return s
+
+    def smiles_to_seq_batch(self, ids_b):
+        """
+        Converts SMILES strings to np.array sequence.
+        """
```
Comment (shreyasvinaya) [inline deepchem/models/torch_models/text_cnn.py:387]: Documentation on parameters to be added
type annotations missing

**Comment by rbharath** [comment]
@Shiva-sankaran Before requesting a new review, make sure to respond to all remaining comments on the PR (can be thumbs up, or a sentence response even) so we know you have addressed all comments

File: deepchem/models/torch_models/text_cnn.py (hunk 2 above)
Comment (Shiva-sankaran) [reply deepchem/models/torch_models/text_cnn.py]: Added parameter descriptions.

File: deepchem/models/torch_models/text_cnn.py (hunk 3 above)
Comment (Shiva-sankaran) [reply deepchem/models/torch_models/text_cnn.py]: Added parameter descriptions.

File: deepchem/models/torch_models/text_cnn.py (hunk 4 above)
Comment (Shiva-sankaran) [reply deepchem/models/torch_models/text_cnn.py:389]: Added parameter descriptions.

File: deepchem/models/torch_models/text_cnn.py (hunk 5 above)
Comment (Shiva-sankaran) [reply deepchem/models/torch_models/text_cnn.py]: Added parameter descriptions and type annotations.

File: deepchem/models/torch_models/text_cnn.py (hunk 6 above)
Comment (Shiva-sankaran) [reply deepchem/models/torch_models/text_cnn.py]: Added parameter descriptions and type annotations.

**Comment by Shiva-sankaran** [comment]
> @Shiva-sankaran Before requesting a new review, make sure to respond to all remaining comments on the PR (can be thumbs up, or a sentence response even) so we know you have addressed all comments

Understood.

File: docs/source/api_reference/models.rst [hunk 7]
```diff
@@ -299,7 +299,7 @@ WGAN
 TextCNNModel
 ------------
 
-.. autoclass:: deepchem.models.TextCNNModel
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:302]: Don't remove the old TF model; add this as a new model

**Review by rbharath — COMMENTED** [review]
One comment to fix on the docs and please rebase the PR

**Comment by rbharath** [comment]
@Shiva-sankaran Please don't merge; rebase instead. Merging messes up large PRs. Can you please rebase?

**Comment by Shiva-sankaran** [comment]
> @Shiva-sankaran Please don't merge; rebase instead. Merging messes up large PRs. Can you please rebase?

Yes, learnt it the hard way. Fixed it.

File: docs/source/api_reference/models.rst (hunk 7 above)
Comment (Shiva-sankaran) [reply docs/source/api_reference/models.rst]: Reverted replacing the old TF model. Added the docs for the torch model under the Torch Models section.

**Review by rbharath — COMMENTED** [review]
@shreyasvinaya if this model looks good to you, can you LGTM it? I will go ahead and merge it in if it looks good to you

File: docs/source/api_reference/molecular_models.csv [hunk 8]
```diff
@@ -11,7 +11,7 @@ MEGNetModel,`ref <https://arxiv.org/abs/1812.05055>`_,Classifier/ Regressor,,PyT
 MPNNModel,`ref <https://arxiv.org/abs/1704.01212>`_,Classifier/ Regressor,MolGraphConvFeaturizer,DGL/PyTorch,
 PagtnModel,`ref <https://arxiv.org/abs/1905.12712>`_,Classifier/ Regressor,PagtnMolGraphFeaturizer MolGraphConvFeaturizer,DGL/PyTorch,
 Smiles2Vec,`ref <https://arxiv.org/abs/1712.02034>`_,Classifier/ Regressor,SmilesToSeq,Keras,
-TextCNNModel,`ref <https://arxiv.org/abs/1705.10843>`_,Classifier/ Regressor,,Keras,
+TextCNNModel,`ref <https://arxiv.org/abs/1705.10843>`_,Classifier/ Regressor,,PyTorch,
```
Comment (shreyasvinaya) [inline docs/source/api_reference/molecular_models.csv:14]: Can you convert this to `Keras/PyTorch` as both implementations exist in the repo

**Comment by shreyasvinaya** [comment]
LGTM, ready to merge once the small docfix is done

File: docs/source/api_reference/molecular_models.csv (hunk 8 above)
Comment (Shiva-sankaran) [reply docs/source/api_reference/molecular_models.csv]: Done.

**Review by shreyasvinaya — APPROVED** [review]
LGTM once CI passes
cc @rbharath

**Comment by rbharath** [comment]
@shreyasvinaya I'm seeing errors in docs-build python 3.11. Are these linked to this PR?

**Comment by shreyasvinaya** [comment]
@Shiva-sankaran can you please rebase the PR, i think the new commits should fix the doctest issues

**Comment by Shiva-sankaran** [comment]
@shreyasvinaya Rebased, there is still 2 doctest failures.

**Comment by shreyasvinaya** [comment]
@rbharath LGTM, good to merge

**Review by rbharath — APPROVED** [review]
LGTM

**Pull request merged**